[4.0] RavenDB-11354 High CPU after hitting leader stepdown several times

karmeli87 opened the pull request:
The issue is related to code for sending topology change notification from the leader.
We waited to leave 'Follower' state in order to reconnect to a new leader, but it could happened that the leader was changed without leaving the follower state, and so we were connected to the wrong node.
This could caused a circular connection between the nodes which sent the same notification again and again.

karmeli87 committed:
RavenDB-11354 High CPU after hitting leader stepdown several times

The issue is related to code for sending topology change notification from the leader.
We waited to leave 'Follower' state in order to reconnect to a new leader, but it could happend that the leader was changed without us leaving the follower state, and so we were connected to the wrong node.
Ihis could caused a circular connection between the nodes which sent the same notification again and again.

ppekrol commented on src/Raven.Server/NotificationCenter/Notifications/Server/ClusterTopologyChanged.cs:
```diff
@@ -53,7 +53,8 @@ public override DynamicJsonValue ToJson()
                 CurrentTerm = term,
-                NodeLicenseDetails = nodeLicenseDetails
+                NodeLicenseDetails = nodeLicenseDetails,
+                IsPersistent = true
```
Why do we want to persist this?

karmeli87 replied:
1. We need to update the topology after reelection even if some nodes are connected after the initial message was sent.
2. Anyway we keep only the recent topology. (the Id defined by the Type in this case)

ayende commented on src/Raven.Server/NotificationCenter/NotificationsBase.cs:
```diff
@@ -17,18 +17,30 @@ public abstract class NotificationsBase : IDisposable
+        {
+            get
+            {
+                lock (_watchersLock)
```
Can we avoid this lock?

ayende commented on src/Raven.Server/NotificationCenter/NotificationsBase.cs:
```diff
@@ -17,18 +17,30 @@ public abstract class NotificationsBase : IDisposable
+            get
             {
-                return _websocketClients > 0 ? Task.CompletedTask : _newWebSocket.Task;
+                lock (_watchersLock)
```
can we avoid this lock?

ayende replied:
We can use `Exchange` to do this atomically, instead

ayende commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -255,7 +255,7 @@ private set
         public event EventHandler LeaderElected;
 
-        private string _tag;
+        private volatile string _tag;
```
Under what scenario is this desirable?

ayende commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -484,6 +484,21 @@ private void SwitchToSingleLeader(TransactionOperationContext context)
+                // we setup the wait _before_ checking the state
+                var task = _stateChanged.Task;
+
+                if (currentLeader != LeaderTag)
```
Better to have the hit here, using `Volatile.Read`, than everywhere.

karmeli87 committed:
RavenDB-11354 fix PR comments

ppekrol committed:
RavenDB-11355
- when null in passed directly to RQL (not as a parameter) then we should detect it properly as NullType, not StringType to match the behavior that we have when null is passed as a parameter
- cleanup

ayende commented on src/Raven.Server/NotificationCenter/NotificationsBase.cs:
```diff
@@ -69,7 +68,7 @@ public IDisposable TrackActions(AsyncQueue<DynamicJsonValue> notificationsQueue,
                 if (watcher.Writer is NotificationCenterWebSocketWriter)
                 {
-                    if (_websocketClients == 0)
+                    if (Interlocked.CompareExchange(ref _haveClients, 1, 0) == 0)
```
You have the wrong order here

ayende committed:
Merge branch 'RavenDB-11354' of https://github.com/karmeli87/ravendb into karmeli87-RavenDB-11354

ayende committed:
RavenDB-11354 - Move to a single variable replacement for concurrent checks

ayende commented:
See how this behaves https://github.com/ravendb/ravendb/pull/6614/commits/c87cbc54d56672ae7d75c9b3626edf20f6a77d61

karmeli87 committed:
RavenDB-11354 Remove unnecessary volatile
